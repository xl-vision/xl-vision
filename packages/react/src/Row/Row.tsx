import React from 'react';
import { styled, ThemeContext } from '../styles';
import { ColProps } from './Col';
import useMedia from './useMedia';

export type RowAlign = 'top' | 'center' | 'bottom';
export type RowJustify = 'start' | 'end' | 'center' | 'space-around' | 'space-between';

export interface RowProps extends React.HTMLAttributes<HTMLDivElement> {
  align?: RowAlign;
  children: React.ReactElement<ColProps> | Array<React.ReactElement<ColProps>>;
  className?: string;
  gutter?: number | Partial<Record<string, number>>;
  justify?: RowJustify;
  type?: 'flex';
}

const RowRoot = styled('div', {
  name: 'Row',
  slot: 'Root',
})<{ align: RowProps['align']; justify: RowProps['justify']; media?: string }>(
  ({ theme, styleProps }) => {
    const { align, justify, media } = styleProps;
    const { mixins } = theme;
    return {
      position: 'relative',
      boxSizing: 'border-box',
      ...mixins.clearfix,
    };
  },
);

const Row: React.FunctionComponent<RowProps> = (props) => {
  const { align = 'center', justify, children, ...others } = props;

  const { breakpoints } = React.useContext(ThemeContext);

  const media = useMedia(breakpoints.values, breakpoints.unit);

  return (
    <RowRoot
      {...others}
      styleProps={{
        align,
        justify,
        media,
      }}
    >
      {children}
    </RowRoot>
  );
};

export default Row;
