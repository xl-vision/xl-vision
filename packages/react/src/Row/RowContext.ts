import React from 'react';

export default React.createContext<{
  matches: Array<string>;
  gutter: number;
}>({
  matches: [],
  gutter: 0,
});
