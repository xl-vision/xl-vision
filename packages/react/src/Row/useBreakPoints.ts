import { useState, useMemo, useEffect } from 'react';
import { useTheme } from '../ThemeProvider';
import { Breakpoint } from '../ThemeProvider/breakpoints';

const useBreakPoints = () => {
  const { breakpoints } = useTheme();

  const { values, unit } = breakpoints;

  const [media, setMedia] = useState<{ [key: string]: boolean }>({});

  const breakPointArray = useMemo(() => {
    return Object.keys(values).sort((a, b) => values[b as Breakpoint] - values[a as Breakpoint]);
  }, [values]);

  useEffect(() => {
    const handlerMap: {
      [breakPoint: string]: {
        listener: EventListener;
        mediaQuery: ReturnType<typeof matchMedia>;
      };
    } = {};
    const keys = Object.keys(values);
    keys.forEach((key) => {
      const size = values[key as Breakpoint];
      const query = `(min-width: ${size}${unit})`;
      const mql = matchMedia(query);
      const onChange = () => {
        setMedia((prev) => ({
          ...prev,
          [key]: mql.matches,
        }));
      };
      handlerMap[key] = {
        listener: onChange,
        mediaQuery: mql,
      };
      mql.addListener(onChange);
      onChange();
    });
    return () => {
      keys.forEach((key) => {
        const { mediaQuery, listener } = handlerMap[key];
        mediaQuery.removeListener(listener);
      });
    };
  }, [values, unit]);

  return useMemo<Array<[Breakpoint, boolean]>>(
    () => breakPointArray.map((key) => [key as Breakpoint, media[key]]),
    [breakPointArray, media],
  );
};

export default useBreakPoints;
