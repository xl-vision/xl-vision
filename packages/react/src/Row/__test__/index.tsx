import { render } from 'enzyme';
import * as React from 'react';
import Row from '..';

describe('Row', () => {
  it('基本用法', () => {
    const wrapper = render(
      <div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={4}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={5}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={7}>
              <div className='col'>col3</div>
            </Row.Col>
            <Row.Col column={8}>
              <div className='col'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={8}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={0}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={8}>
              <div className='col'>col3</div>
            </Row.Col>
            <Row.Col column={8}>
              <div className='col'>col4</div>
            </Row.Col>
          </Row>
        </div>
      </div>,
    );
    expect(wrapper).toMatchSnapshot();
    // wrapper.unmount()
  });
  it('偏移布局', () => {
    const wrapper = render(
      <div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6} offset={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6} push={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6} pull={6}>
              <div className='col'>col3</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={6} offset={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6} pull={12}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col4</div>
            </Row.Col>
          </Row>
        </div>
      </div>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('响应式布局', () => {
    const wrapper = render(
      <div className='box'>
        <Row gutter={{ xs: 8, sm: 10, md: 15, lg: 20, xl: 25, xxl: 30 }}>
          <Row.Col column={{ xs: 8, sm: 6, md: 4, lg: 6 }}>
            <div className='col'>col1</div>
          </Row.Col>
          <Row.Col column={{ xs: 8, sm: 6, md: 5, lg: 10, xxl: 6 }}>
            <div className='col'>col2</div>
          </Row.Col>
          <Row.Col column={{ xs: 8, sm: 6, md: 7, lg: 4, xl: 0, xxl: 6 }}>
            <div className='col'>col3</div>
          </Row.Col>
          <Row.Col column={{ xs: 0, sm: 6, md: 8, lg: 4, xl: 8, xxl: 6 }}>
            <div className='col'>col4</div>
          </Row.Col>
        </Row>
      </div>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('响应式偏移', () => {
    const wrapper = render(
      <div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6} offset={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6} push={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }}>
              <div className='col'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10}>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6} pull={{ xs: 1, sm: 2, md: 3, lg: 4, xl: 5, xxl: 6 }} offset={6}>
              <div className='col'>col3</div>
            </Row.Col>
          </Row>
        </div>
      </div>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('flex布局', () => {
    const wrapper = render(
      <div>
        <div className='box'>
          <Row gutter={10} type='flex' align='top'>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10} type='flex' align='middle'>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10} type='flex' align='bottom'>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col4</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10} type='flex' justify='start'>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10} type='flex' justify='center'>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10} type='flex' justify='end'>
            <Row.Col column={6}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6}>
              <div className='col'>col3</div>
            </Row.Col>
          </Row>
        </div>
        <div className='box'>
          <Row gutter={10} type='flex'>
            <Row.Col column={6} order={3}>
              <div className='col'>col1</div>
            </Row.Col>
            <Row.Col column={6} order={1}>
              <div className='col'>col2</div>
            </Row.Col>
            <Row.Col column={6} order={2}>
              <div className='col'>col3</div>
            </Row.Col>
          </Row>
        </div>
      </div>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
