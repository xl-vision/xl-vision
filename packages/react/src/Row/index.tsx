import Row from './Row';
import Col from './Col';

export { default as Row } from './Row';
export { default as Col } from './Col';

const RowWithCol = Row as typeof Row & {
  Col: typeof Col;
};

RowWithCol.Col = Col;

export default RowWithCol;
