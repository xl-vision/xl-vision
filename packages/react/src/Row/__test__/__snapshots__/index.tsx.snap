// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row flex布局 1`] = `
.c0 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0:before,
.c0:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c0:after {
  clear: both;
}

.c0::after {
  display: none;
}

.c2 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2:before,
.c2:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c2:after {
  clear: both;
}

.c2::after {
  display: none;
}

.c3 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c3:before,
.c3:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c3:after {
  clear: both;
}

.c3::after {
  display: none;
}

.c4 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c4:before,
.c4:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c4:after {
  clear: both;
}

.c4::after {
  display: none;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5:before,
.c5:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c5:after {
  clear: both;
}

.c5::after {
  display: none;
}

.c6 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c6:before,
.c6:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c6:after {
  clear: both;
}

.c6::after {
  display: none;
}

.c7 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c7:before,
.c7:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c7:after {
  clear: both;
}

.c7::after {
  display: none;
}

.c1 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
}

.c8 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
  -webkit-order: 3;
  -ms-flex-order: 3;
  order: 3;
}

.c9 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

.c10 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
}

<div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root xl-row-root-flex xl-row-root-align-top"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c2 xl-row xl-row-root xl-row-root-flex xl-row-root-align-middle"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c3 xl-row xl-row-root xl-row-root-flex xl-row-root-align-bottom"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c4 xl-row xl-row-root xl-row-root-flex xl-row-root-justify-start"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c5 xl-row xl-row-root xl-row-root-flex xl-row-root-justify-center"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c6 xl-row xl-row-root xl-row-root-flex xl-row-root-justify-end"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c7 xl-row xl-row-root xl-row-root-flex"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c8 xl-col xl-col-root xl-col-root-column-6 xl-col-root-order-3"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c9 xl-col xl-col-root xl-col-root-column-6 xl-col-root-order-1"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c10 xl-col xl-col-root xl-col-root-column-6 xl-col-root-order-2"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Row 偏移布局 1`] = `
.c0 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
}

.c0:before,
.c0:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c0:after {
  clear: both;
}

.c1 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
}

.c2 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
  margin-left: 25%;
}

.c3 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
  left: 25%;
}

.c4 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
  right: 25%;
}

.c5 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
  right: 50%;
}

<div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c2 xl-col xl-col-root xl-col-root-column-6 xl-col-root-offset-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c3 xl-col xl-col-root xl-col-root-column-6 xl-col-root-push-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c4 xl-col xl-col-root xl-col-root-column-6 xl-col-root-pull-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c2 xl-col xl-col-root xl-col-root-column-6 xl-col-root-offset-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c5 xl-col xl-col-root xl-col-root-column-6 xl-col-root-pull-12"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Row 响应式偏移 1`] = `
.c0 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
}

.c0:before,
.c0:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c0:after {
  clear: both;
}

.c1 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
}

.c2 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
  margin-left: 25%;
}

<div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c2 xl-col xl-col-root xl-col-root-column-6 xl-col-root-offset-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Row 响应式布局 1`] = `
.c0 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
}

.c0:before,
.c0:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c0:after {
  clear: both;
}

.c1 {
  float: left;
  box-sizing: border-box;
  display: block;
  position: relative;
}

<div
  class="box"
>
  <div
    class="c0 xl-row xl-row-root"
  >
    <div
      class="c1 xl-col xl-col-root"
    >
      <div
        class="col"
      >
        col1
      </div>
    </div>
    <div
      class="c1 xl-col xl-col-root"
    >
      <div
        class="col"
      >
        col2
      </div>
    </div>
    <div
      class="c1 xl-col xl-col-root"
    >
      <div
        class="col"
      >
        col3
      </div>
    </div>
    <div
      class="c1 xl-col xl-col-root"
    >
      <div
        class="col"
      >
        col4
      </div>
    </div>
  </div>
</div>
`;

exports[`Row 基本用法 1`] = `
.c0 {
  position: relative;
  box-sizing: border-box;
  zoom: 1;
}

.c0:before,
.c0:after {
  display: table;
  box-sizing: border-box;
  content: " ";
}

.c0:after {
  clear: both;
}

.c1 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 16.666666666666668%;
}

.c2 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 20.833333333333332%;
}

.c3 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 29.166666666666668%;
}

.c4 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 33.333333333333336%;
}

.c5 {
  float: left;
  box-sizing: border-box;
  position: relative;
  min-height: 1px;
  width: 25%;
}

.c6 {
  float: left;
  box-sizing: border-box;
  display: none;
  position: relative;
  min-height: 1px;
  width: 0%;
}

<div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c1 xl-col xl-col-root xl-col-root-column-4"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c2 xl-col xl-col-root xl-col-root-column-5"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c3 xl-col xl-col-root xl-col-root-column-7"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="c4 xl-col xl-col-root xl-col-root-column-8"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c5 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c5 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c5 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="c5 xl-col xl-col-root xl-col-root-column-6"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
  <div
    class="box"
  >
    <div
      class="c0 xl-row xl-row-root"
      style="margin-left:-5px;margin-right:-5px"
    >
      <div
        class="c4 xl-col xl-col-root xl-col-root-column-8"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col1
        </div>
      </div>
      <div
        class="c6 xl-col xl-col-root"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col2
        </div>
      </div>
      <div
        class="c4 xl-col xl-col-root xl-col-root-column-8"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col3
        </div>
      </div>
      <div
        class="c4 xl-col xl-col-root xl-col-root-column-8"
        style="padding-left:5px;padding-right:5px"
      >
        <div
          class="col"
        >
          col4
        </div>
      </div>
    </div>
  </div>
</div>
`;
